etcdserver/server.go

```diff
@@ -854,6 +860,10 @@ func (s *EtcdServer) apply(es []raftpb.Entry, confState *raftpb.ConfState) (uint
 					req := raftReq.V2
 					s.w.Trigger(req.ID, s.applyRequest(*req))
 				default:
+					// reject to apply entry that has been executed before
+					if e.Index <= s.kv.ConsistentIndex() {
```

Comment:
after trying it, i still think it is apply loop's responsibility to avoid double apply.

One reason is that we want kv to be as dumb as possible. For now, it just adds the ability to store/load consistent index. If it handles apply duplication, it needs error returned from `TxnBegin`, and may take over more apply logic to become complicated in the future.

Another reason is that we have worked on truncating log entries before applying them at https://github.com/coreos/etcd/pull/3746/files#diff-971f6af906b31cc86efd71aff56aa96bR544 . This could replace if statement here when we removing support of v2.